source/opt/scalar_analysis.cpp

```diff
@@ -0,0 +1,386 @@
+  }
+  add_node->AddChild(op2);
+
+  return GetCachedOrAdd(std::move(add_node));
+}
+
+SENode* ScalarEvolutionAnalysis::AnalyzePhiInstruction(
+    const ir::Instruction* phi) {
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+
+  // Get the basic block this instruction belongs to.
+  ir::BasicBlock* basic_block =
+      context_->get_instr_block(const_cast<ir::Instruction*>(phi));
+
+  // And then the function that the basic blocks belongs to.
+  ir::Function* function = basic_block->GetParent();
+
+  // Use the function to get the loop descriptor.
+  ir::LoopDescriptor* loop_descriptor = context_->GetLoopDescriptor(function);
+
+  // We only handle phis in loops at the moment.
+  if (!loop_descriptor) return CreateCantComputeNode();
+
+  // Get the innermost loop which this block belongs to.
+  ir::Loop* loop = (*loop_descriptor)[basic_block->id()];
+  if (!loop) return CreateCantComputeNode();
+
+  std::unique_ptr<SERecurrentNode> phi_node{new SERecurrentNode(this, loop)};
+  instruction_map_[phi] = phi_node.get();
+
+  for (uint32_t i = 0; i < phi->NumInOperands(); i += 2) {
+    uint32_t value_id = phi->GetSingleWordInOperand(i);
+    uint32_t incoming_label_id = phi->GetSingleWordInOperand(i + 1);
+
+    ir::Instruction* value_inst = def_use->GetDef(value_id);
+    SENode* value_node = AnalyzeInstruction(value_inst);
+
+    // Both operands must be loop invariant.
+    if (!IsLoopInvariant(loop, value_node)) {
+      return CreateCantComputeNode();
+    }
+
+    if (incoming_label_id == loop->GetPreHeaderBlock()->id()) {
+      phi_node->AddOffset(value_node);
+    } else if (incoming_label_id == loop->GetLatchBlock()->id()) {
+      // Assumed to be in the form of step + phi.
+      SENode* constant_node = nullptr;
+      SENode* phi_operand = nullptr;
+      SENode* operand_1 = value_node->GetChild(0);
+      SENode* operand_2 = value_node->GetChild(1);
+
+      // Find which node is the step term.
+      if (!operand_1->AsSERecurrentNode())
+        constant_node = operand_1;
+      else if (!operand_2->AsSERecurrentNode())
+        constant_node = operand_2;
+
+      // Find which node is the recurrent expression.
+      if (operand_1->AsSERecurrentNode())
+        phi_operand = operand_1;
+      else if (operand_2->AsSERecurrentNode())
+        phi_operand = operand_2;
+
+      // If it is not in the form step + phi exit out.
+      if (!(constant_node && phi_operand)) return CreateCantComputeNode();
+
+      // If the phi operand is not the same phi node exit out.
+      if (phi_operand != phi_node.get()) return CreateCantComputeNode();
+
+      phi_node->AddCoefficient(constant_node);
+    }
+  }
+
+  instruction_map_[phi] = GetCachedOrAdd(std::move(phi_node));
+
+  return instruction_map_[phi];
+}
+
+SENode* ScalarEvolutionAnalysis::CreateValueUnknownNode(
+    const ir::Instruction* inst) {
+  std::unique_ptr<SEValueUnknown> load_node{
+      new SEValueUnknown(this, inst->unique_id())};
```

Comment:
I would prefer if we used the result_id instead of the unique_id.  If at some point we need a link back to the original instruction, then we can get the instruction easily from the id.